### ISSUE TITLE
possible overwriting of backups of unsaved tabs with non-unique filenames from "-multiInst" instances

### ISSUE THREAD

**Issue by xomx** [notepad-plus-plus/notepad-plus-plus#14691]
Description:
https://community.notepad-plus-plus.org/topic/25434/bug-npp-loses-all-open-files-after-restarting-as-admin/17

The following is due to the "new x" tab filenames are not multi-instance unique.
The problem is only with the cmdline parameter "-multiInst", not with such setting in Preferences (which effectively disables possible N++ session backup feature in any non-first instance).

STR (using x64 v8.6.2 portable):

1)
- check Preferences > Backup > Remember current session for next launch & Enable session snapshot and periodic backup
- note how many seconds you have there in the Backup in every XXX seconds (default here is 7, I will use this number in my STR)
- set Preferences > Multi-Instance & Date > Muti-instance settings == Always in multi-instance mode
- close N++
- launch 1st instance of N++, there will be the "new 1" tab, edit its content e.g. by adding number "1"
- wait 7 secs, then check the backup created (a  .\\backup\\new 1@YYYY-MM-DD_HHMMSS) and its content (should be that "1")
- launch 2nd N++ instance, there will be the same "new 1" tab,  edit its content e.g. by adding number "2"
- wait another 7 secs, then check the original backup file (the same  .\\backup\\new 1@YYYY-MM-DD_HHMMSS)), its content remains that "1"

2)
- change the Preferences > Multi-Instance & Date > Muti-instance settings == Default (mono-instance)
- close N++
- repeat the remaining test as in the above first case, but this time launch all the instances of N++ with "-multiInst" cmdline param
- result: 2nd N++ instance will overwrite the backup of the 1st N++ instance with "2"

**Comment by donho** [comment]
@xomx 
What's your expected result for both scenarios?

**Comment by xomx** [comment]
@donho 
The behavior of the 1st one (multi-inst set by Preferences). So no overwriting of the same backup from the multi-inst N++ instances.

The N++ multi-inst mode needs some revision or at least some additional notes/warnings about it to the N++ manual (I already asked for such [warning note](https://community.notepad-plus-plus.org/post/92666) for the multi-inst vs backup). But there is another example of such unexpected behavior - using multi-inst & admin-mode: #14694 .

**Comment by donho** [comment]
I agree with you that incoherent behaviour should be fixed.
However, applying `-nosession` automatically while `-multiInst` being used is not a solution to me (not mentioning the behaviour's change will cause a lot complains from the users):
The "solution" cannot prevent the 2nd instance of Notepad++ from opening the files opened in the 1st instance, then modifies them alternately in both instances - the same result of your the 2nd scenario will happen (2nd N++ instance will overwrite the backup of the 1st N++ instance).

The "real" solution will be, IMO, allowing the same files to be opened in 2, 3 or more instances, and synchronize the status/content of the files in questions among these instances.

What do you think?

**Comment by rdipardo** [comment]
> allowing the same files to be opened in 2, 3 or more instances, and synchronize the status/content of the files in questions among these instances.

# 🤯

**Comment by xomx** [comment]
@donho 

> However, applying `-nosession` automatically while `-multiInst` being used (#14692) is not a solution to me (not mentioning the behaviour's change will cause a lot complains from the users)

My intention was not change any behavior, I just tried to make the N++ multi-inst modes coherent in behavior.
But I probably made the mistake of forgetting that there is a third N++ Preferences mode that is(?) the proper equivalent of the cmdline "-multiInst":
![npp-Preferences-3rd-multiinstmode](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/13075183/8d0ae4cf-6668-4157-ae2b-8c8054b4bc81)
Is that so?

> The "solution" cannot prevent the 2nd instance of Notepad++ from opening the files opened in the 1st instance by the user, then modifies them alternately in both instances - the same result of your the 2nd scenario will happen (2nd N++ instance will overwrite the backup of the 1st N++ instance).

Are you still talking about the backup-files only? If so I agree with you (and say - good point!)

> allowing the same files to be opened in 2, 3 or more instances, and synchronize the status/content of the files in questions among these instances.

You mean allowing this N++ mechanism also for these backup-files?
![npp-cross-process-file](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/13075183/28bf3513-77bf-41dc-94e7-99243513c812)

> What do you think?

For the N++ backup mechanism, I can think of one simpler improvement that would help to eliminate the non-uniqueness of filenames used for such backups:
- instead of using the current  .\\backup\\`tab/filename@YYYY-MM-DD_HHMMSS`
- use the  .\\backup\\`tab/filename@YYYY-MM-DD_HHMMSS_PID` (Windows process ID)
  (1st considered here the N++ BufferID for it but there is still a very small but non-zero probability of the same BufferID in between the different N++ instances because of every Windows app has its own virtual memory space...)

**Comment by donho** [comment]
@xomx 
> My intention was not change any behavior, I just tried to make the N++ multi-inst modes coherent in behavior.
But I probably made the mistake of forgetting that there is a third N++ Preferences mode that is(?) the proper equivalent of the cmdline "-multiInst":*

If you are talking only about making command line argument `-multiInst` coherent, then there's nothing to do about it - this command line argument does what we expect exactly. 
But if you're talking about 3 options of "Multi-instance settings", then they have nothing to do with `-multiInst`. Some of them may use `-multiInst` to get the job done in the code, but from the aspect of users their different behaviours are disassociated with `-multiInst`. So nothing to be done either here.

>>  The "solution" cannot prevent the 2nd instance of Notepad++ from opening the files opened in the 1st instance by the user, then modifies them alternately in both instances - the same result of your the 2nd scenario will happen (2nd N++ instance will overwrite the backup of the 1st N++ instance).

> Are you still talking about the backup-files only? If so I agree with you (and say - good point!)

Yes, I am talking about overwriting of backup files.

> You mean allowing this N++ mechanism also for these backup-files?

This is what I am talking about:

![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/90293/5c3221a8-2375-46de-a495-e221433e8f3b)

`pack.log` file has been modified but not saved in the previous session, and its content is "1". The multi-instance settings of instance 1 is default mono-instance. Then we launch the instance 2 by using `-muliInst` as argument. The instance 2 opened the same session, and it had the same content "1". We modified in the content of instance 2 from "1" to "2" - that's the state of above screenshot.

What I was trying to say is: when we change the focus from the instance 2 to instance 1 (without saving content "2" in instance 2), the content "1" in instance 1 should become "2", and the content state of instance 1 remains dirty surely - that's the synchronization I meant.


> use the .\backup\tab/filename@YYYY-MM-DD_HHMMSS_PID (Windows process ID)

It won't work, because as I know process ID changes all the time, so it could solve the multi-instance problem we have here, but next launch time backup files won't be loaded due to the PID's change.

**Comment by xomx** [comment]
@donho 

> > use the .\backup\tab/filename@YYYY-MM-DD_HHMMSS_PID (Windows process ID)
> 
> It won't work, because as I know process ID changes all the time, so it could solve the multi-instance problem we have here, but next launch time backup files won't be loaded due to the PID's change.

Yes, that idea with _PID will not work at the next launch time, but due to the fact that a 2nd concurrent instance of the N++ launched via the cmdline "-multiInst" will have its `_isNoSession == false` (unlike the N++ GUI Preferences Always in multi-instance mode setting, which will will have its `_isNoSession == true` and thus will be fine), so it will load the same session.xml as the 1st N++ instance where it will read the same `backupFilePath` item value the 1st N++ instance is already using.

I need more time to think about it and do some tests.

**Comment by donho** [comment]
@xomx 
I think we should focus on "synchronizing" of backup files' overwritten instead of preventing loading session of `-multiInst`.
As I've said, even we make `-multiInst` argument of the 2nd instance not loading the session, user can always open the files which are already opened in the 1st instance - then backup files' overwritten problem still happens.
